File: packages/java/src/test/java/ai/alumnium/unit/tool/ToolToSchemaConverterTest.java
package ai.alumnium.unit.tool;

import static org.assertj.core.api.Assertions.assertThat;

import ai.alumnium.tool.ClickTool;
import ai.alumnium.tool.ToolToSchemaConverter;
import ai.alumnium.tool.TypeTool;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class ToolToSchemaConverterTest {

  @Test
  void convertsClickToolToFunctionSchema() {
    Map<String, Object> schema = ToolToSchemaConverter.convert(ClickTool.class);

    assertThat(schema).containsEntry("type", "function");

    @SuppressWarnings("unchecked")
    Map<String, Object> function = (Map<String, Object>) schema.get("function");
    assertThat(function)
        .containsEntry("name", "ClickTool")
        .containsEntry(
            "description",
            "Click an element. If the target element is a dropdown and is already expanded"
                + " - you don't need to click it. NEVER use ClickTool to upload files - use"
                + " UploadTool instead.");

    @SuppressWarnings("unchecked")
    Map<String, Object> parameters = (Map<String, Object>) function.get("parameters");
    assertThat(parameters).containsEntry("type", "object");

    @SuppressWarnings("unchecked")
    Map<String, Object> properties = (Map<String, Object>) parameters.get("properties");
    @SuppressWarnings("unchecked")
    Map<String, Object> idProp = (Map<String, Object>) properties.get("id");
    assertThat(idProp)
        .containsEntry("type", "integer")
        .containsEntry("description", "Element identifier (ID)");

    @SuppressWarnings("unchecked")
    List<String> required = (List<String>) parameters.get("required");
    assertThat(required).containsExactly("id");
  }

  @Test
  void convertsTypeToolToFunctionSchema() {
    Map<String, Object> schema = ToolToSchemaConverter.convert(TypeTool.class);

    assertThat(schema).containsEntry("type", "function");

    @SuppressWarnings("unchecked")
    Map<String, Object> function = (Map<String, Object>) schema.get("function");
    assertThat(function)
        .containsEntry("name", "TypeTool")
        .containsEntry(
            "description",
            "Type text into an element. Automatically focuses the element and clears it"
                + " before typing.");

    @SuppressWarnings("unchecked")
    Map<String, Object> parameters = (Map<String, Object>) function.get("parameters");
    assertThat(parameters).containsEntry("type", "object");

    @SuppressWarnings("unchecked")
    Map<String, Object> properties = (Map<String, Object>) parameters.get("properties");
    @SuppressWarnings("unchecked")
    Map<String, Object> idProp = (Map<String, Object>) properties.get("id");
    assertThat(idProp)
        .containsEntry("type", "integer")
        .containsEntry("description", "Element identifier (ID)");

    @SuppressWarnings("unchecked")
    List<String> required = (List<String>) parameters.get("required");
    assertThat(required).containsExactly("id", "text");
  }

  @Test
  void convertAllReturnsOneSchemaPerTool() {
    List<Map<String, Object>> schemas =
        ToolToSchemaConverter.convertAll(
            Map.of("ClickTool", ClickTool.class, "TypeTool", TypeTool.class));

    assertThat(schemas).hasSize(2);
    schemas.forEach(
        (tool) -> {
          assertThat(tool).containsEntry("type", "function");
        });
  }
}

File: packages/typescript/src/tools/toolToSchemaConverter.test.ts
import { describe, expect, it } from "vitest";
import { PressKeyTool } from "./PressKeyTool.ts";
import { convertToolsToSchemas } from "./toolToSchemaConverter.ts";
import { TypeTool } from "./TypeTool.ts";
import { UploadTool } from "./UploadTool.ts";

describe(convertToolsToSchemas, () => {
  it("converts tool with primitives", () => {
    expect(convertToolsToSchemas({ TypeTool })[0]).toEqual({
      type: "function",
      function: {
        name: "TypeTool",
        description:
          "Type text into an element. Automatically focuses the element and clears it before typing.",
        parameters: {
          type: "object",
          properties: {
            id: {
              type: "integer",
              description: "Element identifier (ID)",
            },
            text: {
              type: "string",
              description: "Text to type into an element",
            },
          },
          required: ["id", "text"],
        },
      },
    });
  });

  it("converts tool with enum", () => {
    const schema = convertToolsToSchemas({ PressKeyTool })[0];

    expect(schema).toEqual({
      type: "function",
      function: {
        name: "PressKeyTool",
        description:
          "Press a keyboard key. Does not require element to be focused.",
        parameters: {
          type: "object",
          properties: {
            key: {
              type: "string",
              enum: ["Backspace", "Enter", "Escape", "Tab"],
              description: "Key to press.",
            },
          },
          required: ["key"],
        },
      },
    });
  });

  it("converts tool with array", () => {
    const schema = convertToolsToSchemas({ UploadTool })[0];

    expect(schema).toEqual({
      type: "function",
      function: {
        name: "UploadTool",
        description:
          "Upload one or more files using a button that opens a file chooser. " +
          "This tool automatically clicks the button, DO NOT use ClickTool for that.",
        parameters: {
          type: "object",
          properties: {
            id: {
              type: "integer",
              description: "Element identifier (ID)",
            },
            paths: {
              type: "array",
              items: {
                type: "string",
              },
              description:
                "Absolute file path(s) to upload. Can be a single path or multiple paths for multi-file upload.",
            },
          },
          required: ["id", "paths"],
        },
      },
    });
  });
});
